src/Nerdbank.MessagePack/StandardVisitor.cs

```diff
@@ -174,7 +172,65 @@ internal StandardVisitor(ConverterCache owner, TypeGenerationContext context)
 			}
 		}
 
-		return unionTypes is null ? converter : new SubTypeUnionConverter<T>(unionTypes, converter);
+		return this.DiscoverUnionTypes(objectShape, converter) is { } unionTypes ? new UnionConverter<T>(converter, unionTypes) : converter;
+	}
+
+	/// <inheritdoc/>
+	public override object? VisitUnion<TUnion>(IUnionTypeShape<TUnion> unionShape, object? state = null)
+	{
+		MessagePackConverter<TUnion> baseTypeConverter = (MessagePackConverter<TUnion>)unionShape.BaseType.Accept(this)!;
+
+		if (baseTypeConverter is UnionConverter<TUnion>)
+		{
+			// A runtime mapping *and* attributes are defined for the same base type.
+			// The runtime mapping has already been applied and that trumps attributes.
+			// Just return the union converter we created for the runtime mapping to avoid
+			// double-nesting.
+			return baseTypeConverter;
+		}
+
+		// Runtime mapping overrides attributes.
+		if (!(unionShape.BaseType is IObjectTypeShape<TUnion> baseObjectShape && this.DiscoverUnionTypes(baseObjectShape, baseTypeConverter) is { } subTypes))
+		{
+			Getter<TUnion, int> getUnionCaseIndex = unionShape.GetGetUnionCaseIndex();
+			Dictionary<int, MessagePackConverter> deserializerByIntAlias = new(unionShape.UnionCases.Count);
+			List<(DerivedTypeIdentifier Alias, MessagePackConverter Converter, ITypeShape Shape)> serializers = new(unionShape.UnionCases.Count);
+			KeyValuePair<int, MessagePackConverter<TUnion>>[] unionCases = unionShape.UnionCases
+				.Select(unionCase =>
+				{
+					// TODO: Use the Name if Tag isn't set explicitly for better schema stability.
+					// Poor man's substitute: if Tag != Index, we know the user set it explicitly.
+					// But they might have set it explicitly to be the same as Index, so we need a feature from PolyType.
+					DerivedTypeIdentifier alias = unionCase.Tag != unionCase.Index ? new(unionCase.Tag) : new(unionCase.Name);
```

Comment:
This needs an `IUnionCaseShape.IsTagSpecified` property or similar so that we can use the tag _iff_ the user explicitly specified one. Using an implicitly specified one is no good unless the user has opted into an "unstable schema" mode (which we may introduce for https://github.com/aarnott/nerdbank.MessagePack/issues/267).

Reply:
If you want to play it safe, you could also just resolve the attribute at runtime and check if the value equals -1:

https://github.com/eiriktsarpalis/PolyType/blob/main/src/PolyType/DerivedTypeShapeAttribute.cs#L44

Reply:
That's reasonable. It requires I search all the attributes for a match, but it works.
https://github.com/AArnott/Nerdbank.MessagePack/compare/74bc99f76125f3954b3e5f3e4e93e87c8516b9e1..e93dc61de1b9ecb8b96719e3ef6db2977d6161fa